### ISSUE TITLE
Move node to device?

### ISSUE THREAD

**Issue by o1lo01ol1o** [tensorflow/haskell#89]
If I have the the following expression:

```haskell
[...]
let deadSoon = x ``TF.add`` y
let out = deadSoon ``TF.add`` z
[...]
```
Is it possible to move `deadSoon` to another device after the computation of `out`?  If so, and this is a forward pass, will back propagation of the graph move `deadSoon`or will there be some sort of temp device transfer each time `deadSoon` is needed?

**Comment by blackgnezdo** [comment]
I doubt the TF operational model is specified with enough precision to guarantee when some transfers will happen. I believe that currently the nodes are placed once and it is possible to [assign where they will be placed](https://github.com/tensorflow/haskell/blob/2c5c879037726001053a02cdfbca62642ac858e5/tensorflow/src/TensorFlow/Build.hs#L339) using TF Haskell APIs.

Are you trying to build a mental model of TF performance or facing a particular problem?

**Comment by o1lo01ol1o** [comment]
A little of both.  In cases of very long sequences over which one would like to fold (say, RNNs), the python API moves those tensors that are dead on arrival (within the recurrent context) off the gpu (see [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/control_flow_ops.py#L2547)).  So I was trying to understand to what extent that is possible in the current haskell api.  From your answer, I assume that `doa` in
```haskell
[. . .]
let doa = x ``TF.add`` y
in x
[. . .]
```
could be assigned to the cpu when it is bound to `let`.

**Comment by blackgnezdo** [comment]
@o1lo01ol1o presumably you are interested in swap_memory attribute which seems to be defined by  StackPushOp. The attribute should be settable just like any other, but all the high level WhileLoop code does not exist in Haskell.

As for your particular example, doa would be built into a graph and where it is placed depends on whether or not you use withDevice. Otherwise you are at the mercy of graph placer code.

**Comment by o1lo01ol1o** [comment]
@blackgnezdo Does the current gradient accumulation code (ie, for backprop) support merge, enter, and exit ops as used in forward propagation? Or does all of the context tracking defined in that python module need to be implemented and then parsed in the gradient module?

**Comment by blackgnezdo** [comment]
I'm afraid the latter is the case. You can find [all the ops that we support in Gradient.hs](https://github.com/tensorflow/haskell/blob/c99a23b6a746949136ae193d3982de5ee2a4c414/tensorflow-ops/src/TensorFlow/Gradient.hs#L670). There's a long-term goal of using the gradient logic defined in C++ once it is more complete than what we have. But it will be a big an invasive change. Possibly even causing breakage of the current API for defining graphs.

**Comment by o1lo01ol1o** [comment]
I see.  So until that happens, recurrent architectures won't be possible unless someone reimplements the entirety of the control flow module as well as the necessary grad ops.  Is there a roadmap for that?

**Comment by blackgnezdo** [comment]
Nobody I know is working on this as their day job. It will come down to somebody having an itch to scratch. Would you like to be that somebody? :)

**Comment by o1lo01ol1o** [comment]
I have an itch but with that scope and my current budget on one-to-two-hour-a-week projects, I'm only slightly optimistic that I'll be able to contribute something useful.    I'll see how far it gets.  :)

Closing this issue.

**Comment by fkm3** [comment]
You can do an RNN [static_rnn](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/static_rnn) style. I should have an example of this ready in the near future. I'm not sure how much work supporting gradients for control flow will be.

**Comment by fkm3** [comment]
Although be warned that the gradient code goes exponential right now so that isn't feasible without some tweaks to Gradient.hs. https://github.com/tensorflow/haskell/pull/88 will fix that issue.
